packages/preset-mini/src/utils/utilities.ts

```diff
@@ -7,7 +9,177 @@ export function capitalize<T extends string>(str: T) {
+        colorData = getThemeColor(theme, [name])
+      }
+    }
+
+    if (typeof colorData === 'string')
+      color = colorData
+    else if (no && colorData)
+      color = colorData[no]
+  }
+
+  return {
+    opacity,
+    name,
+    no,
+    color,
+    rgba: hex2rgba(color),
+  }
+}
+
+export const colorResolver = (attribute: string, varName: string) => ([, body]: string[], { theme }: RuleContext<Theme>) => {
+  const data = parseColor(body, theme)
+
+  if (!data)
+    return
+
+  const { opacity, color, rgba } = data
+
+  if (!color)
+    return
+
+  const a = opacity
+    ? opacity[0] === '['
+      ? h.bracket.percent(opacity)!
+      : (parseFloat(opacity) / 100)
+    : rgba?.[3]
+
+  if (rgba) {
+    if (a != null && !Number.isNaN(a)) {
+      // @ts-expect-error
+      rgba[3] = typeof a === 'string' && !a.includes('%')
+        ? parseFloat(a)
+        : a
+      return {
+        [attribute]: `rgba(${rgba.join(',')})`,
+      }
+    }
+    else {
+      return {
+        [`--un-${varName}-opacity`]: 1,
+        [attribute]: `rgba(${rgba.slice(0, 3).join(',')},var(--un-${varName}-opacity))`,
+      }
+    }
+  }
+  else {
+    return {
+      [attribute]: color.replace('%alpha', `${a || 1}`),
+    }
+  }
+}
+
+export const colorOpacityResolver = (varName: string) => ([, o]: string[]) => {
+  const v = h.bracket.percent.cssvar(o)
+  if (v !== undefined)
+    return { [`--un-${[varName]}-opacity`]: v }
+}
+
+export const createColorRule = (name: string, prop?: string, varName?: string): Rule => [
+  new RegExp(`^${escapeRegExp(name)}-(.+)$`),
+  colorResolver(prop ?? name, varName ?? name),
+]
+
+export const createColorOpacityRule = (name: string, varName?: string): Rule => [
+  new RegExp(`^${escapeRegExp(name)}-op(?:acity)?-?(.+)$`),
+  colorOpacityResolver(varName ?? name),
+]
+
+export const createColorAndOpacityRulePair = (name: string, prop?: string, varName?: string) => [
+  createColorRule(name, prop, varName),
+  createColorOpacityRule(name, varName),
+]
+
+function generateCssObject(props: string | string[], value: string | number) {
+  if (typeof props === 'string')
+    return { [props]: value }
+
+  const css: CSSObject = {}
+  props.forEach(prop => css[prop] = value)
+  return css
+}
+
+export const keywordResolver = (props: string | string[], keywords: CSSPropertyValue[] = []) => {
+  const cache = Object.fromEntries(keywords.map(toArray))
+  return ([, s]: string[]) => {
+    if (s in cache)
+      return generateCssObject(props, cache[s] ?? s)
+  }
+}
+
+export const globalKeywordResolver = (props: string | string[], keywords: CSSPropertyValue[] = []) => {
+  const cache = Object.fromEntries(keywords.map(toArray))
+  return ([, s]: string[]) => {
+    const v = s in cache ? (cache[s] ?? s) : h.global(s)
+    if (v !== undefined)
+      return generateCssObject(props, v)
+  }
+}
+
+export const createKeywordRules = (bases: string | string[], props: string | string[], keywords: CSSPropertyValue[] = []): Rule[] =>
```

Comment:
Overall looks good, but the abstraction makes the code hard to read. Can you write some jsdocs for each utilities and provide some input / output examples to them so we could better understand how each functions doing their jobs? Thanks